rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,117 @@
+* (+) Enables communicating language changes systematically.
+* (+) Backwards-compatible
+  * (+) Allows for gradual adoption: opt-in until semantics is implemented in Nix *and* the first backwards-incompatible change to the language is introduced.
+* (+) Visually unintrusive
+* (+) Self-describing and human-readable
+* (+) Follows a well-known convention of using [magic numbers in files](https://en.m.wikipedia.org/wiki/Magic_number_(programming)#In_files)
+* (-) May make the appearance that changing the language is harmless.
+  * (+) The convention itself is harmless and independent of the development culture around the language.
+* (-) The syntax of the magic comment is arbitrary.
+* (-) There is a chance of abusing the magic comment for more metadata in the future. Let's avoid that.
+* (-) At least one form of comment is forever bound to begin with `#` to maintain compatibility.
+* (-) It requires significant additional effort to implement and maintain an appropriate system to make use of the version information.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Never introduce backwards-incompatible changes to the language.
+
+  * (+) No additional effort required.
+  * (-) Requires additions to be made very carefully.
+  * (-) Makes solving some well-known problems impossible.
+
+- Use the output of [`builtins.langVersion`] for specifying the version of the Nix language.
+
+  * (+) This would serve other Nix language evaluators which are not and should not be tied to the rest of Nix.
+  * (-) `builtins.langVersion` is currently only internal and undocumented.
+    * (+) Documentation is easy to add.
+      * (-) Requires adding another built-in to the public API.
+    * (-) Using a language feature requires an additional steps from users to determine the current version.
+      * (+) We can add a command line option such that it is not more effort than `nix --version`.
+        * (-) Requires adding another command line option to the public API.
+  * (+) The Nix language version is decoupled Nix version numbering.
+    * (+) It changes less often than the Nix version.
+      * (-) That was probably due to making changes being so hard.
+    * (-) There are two version numbers to keep track of.
+  * (-) The magic comment should reflect that it's specifying the *Nix language* version, which would make it longer.
+    * (-) Renaming the Nix language will be impossible once the mechanism is part of stable Nix.
+
+[`builtins.langVersion`]: https://github.com/NixOS/nix/blob/26c7602c390f8c511f326785b570918b2f468892/src/libexpr/primops.cc#L3952-L3957
+
+- Use a magic string that is incompatible with evaluators prior to the feature, e.g. `%? Nix <version>`.
+
+  * (+) Makes clear that the file is not intended to be used without explicit handling of compatibility.
+    * (-) Cannot be introduced gradually.
+  * (+) Such a breaking change could also be reserved for later iterations of the Nix language.
+
+# Prior art
```

Comment:
Stumbled upon https://github.com/hsutter/cppfront, I think it should be discussed here too. It does not do the same thing as this RFC proposes, but there is some overlap in motivation and provided features.

Reply:
On an inspired day some time ago, I sketched https://md.darmstadt.ccc.de/nix2. I think it shows that one can get pretty far with syntax changes alone, without having to touch any of the evaluation semantics of Nix.